bb-junit5/src/main/java/com/cognifide/qa/bb/junit5/guice/GuiceExtension.java

```diff
@@ -0,0 +1,198 @@
+      return getOrCreateInjector(optionalParent.get());
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Gets all new {@link Module} instances for injections
+   */
+  private static List<? extends Module> getNewModules(ExtensionContext context)
+      throws NoSuchMethodException, InstantiationException, IllegalAccessException, InvocationTargetException {
+    Set<Class<? extends Module>> moduleTypes = getNewModuleTypes(context);
+    List<Module> modules = new ArrayList<>(moduleTypes.size());
+    for (Class<? extends Module> moduleType : moduleTypes) {
+      Constructor<? extends Module> moduleCtor = moduleType.getDeclaredConstructor();
+      moduleCtor.setAccessible(true);
+
+      modules.add(moduleCtor.newInstance());
+    }
+
+    context.getElement().ifPresent(element -> {
+      if (element instanceof Class) {
+        modules.add(new AbstractModule() {
+          @Override
+          protected void configure() {
+            requestStaticInjection((Class<?>) element);
+          }
+        });
+      }
+    });
+
+    return modules;
+  }
+
+  /**
+   * Returns all new {@link Module} declared in current context (returns empty set if modules where
+   * already declared)
+   */
+  private static Set<Class<? extends Module>> getNewModuleTypes(ExtensionContext context) {
+    Optional<AnnotatedElement> optionalAnnotatedElement = context.getElement();
+    if (!optionalAnnotatedElement.isPresent()) {
+      return Collections.emptySet();
+    }
+
+    Set<Class<? extends Module>> moduleTypes = getModuleTypes(optionalAnnotatedElement.get());
+    context.getParent()
+        .map(GuiceExtension::getContextModuleTypes)
+        .ifPresent(moduleTypes::removeAll);
+
+    return moduleTypes;
+  }
+
+  private static Set<Class<? extends Module>> getContextModuleTypes(ExtensionContext context) {
+    return getContextModuleTypes(Optional.of(context));
+  }
+
+  /**
+   * Returns module types that are present on the given context or any of its enclosing contexts.
+   */
+  private static Set<Class<? extends Module>> getContextModuleTypes(
+      Optional<ExtensionContext> context) {
+
+    Set<Class<? extends Module>> contextModuleTypes = new LinkedHashSet<>();
+    while (context.isPresent() && (hasAnnotatedElement(context) || hasParent(context))) {
+      context
+          .flatMap(ExtensionContext::getElement)
+          .map(GuiceExtension::getModuleTypes)
+          .ifPresent(contextModuleTypes::addAll);
+      context = context.flatMap(ExtensionContext::getParent);
+    }
+
+    return contextModuleTypes;
+  }
+
+  private static boolean hasAnnotatedElement(Optional<ExtensionContext> context) {
+    return context.flatMap(ExtensionContext::getElement).isPresent();
+  }
+
+  private static boolean hasParent(Optional<ExtensionContext> context) {
+    return context.flatMap(ExtensionContext::getParent).isPresent();
+  }
+
+  private static Set<Class<? extends Module>> getModuleTypes(AnnotatedElement element) {
```

Comment:
Could be rewritten using the `stream.orElse()` method.

Reply:
It has problems with generics array